packages/service/src/service/offline-service.ts

```diff
@@ -0,0 +1,168 @@
+import { ErrorCodes } from '../error';
+import { BtcAssetsApi } from './service';
+import { OfflineBtcAssetsDataSourceError } from '../error';
+
+import {
+  BtcApiTransaction,
+  BtcApiUtxo,
+  BtcApiUtxoParams,
+  BtcApiTransactionParams,
+  BtcApiBalanceParams,
+  RgbppApiTransactionStateParams,
+  RgbppApiAssetsByAddressParams,
+  RgbppApiBalanceByAddressParams,
+  RgbppApiActivityByAddressParams,
+  RgbppApiSendCkbTransactionPayload,
+  RgbppApiRetryCkbTransactionPayload,
+  OfflineBtcUtxo,
+} from '../types';
+
+export interface OfflineBtcData {
+  txs: BtcApiTransaction[];
+  utxos: OfflineBtcUtxo[];
+}
+
+/*
+ * Currently, the offline mode is only suitable for scenarios where you prefer to manually provide transactions and UTXOs data to build BTC transactions,
+ * as its implemented methods are limited. If you require access to the full range of functionalities, such as advanced queries or sending transactions,
+ * consider switching to BtcAssetsApi instead.
+ */
+export class OfflineBtcAssetsDataSource extends BtcAssetsApi {
+  // txid -> tx
+  private txs: Record<string, BtcApiTransaction>;
+  // address -> utxos
+  private utxos: Record<string, OfflineBtcUtxo[]>;
+
+  constructor(offlineData: OfflineBtcData) {
+    super({ url: 'DUMMY_URL' });
+
+    this.txs = offlineData.txs.reduce(
+      (acc, tx) => {
+        acc[tx.txid] = tx;
+        return acc;
+      },
+      {} as Record<string, BtcApiTransaction>,
+    );
+
+    this.utxos = offlineData.utxos.reduce(
+      (acc, utxo) => {
+        acc[utxo.address] = [...(acc[utxo.address] ?? []), utxo];
+        return acc;
+      },
+      {} as Record<string, OfflineBtcUtxo[]>,
+    );
+  }
+
+  getBtcTransaction(txId: string): Promise<BtcApiTransaction> {
+    return Promise.resolve(this.txs[txId]);
+  }
+
+  getBtcUtxos(address: string, params?: BtcApiUtxoParams): Promise<BtcApiUtxo[]> {
+    const addressUtxos = this.utxos[address] || [];
+    const { only_non_rgbpp_utxos, only_confirmed, min_satoshi } = params || {};
+
+    return Promise.resolve(
+      addressUtxos.filter(
+        (utxo) =>
+          (!only_non_rgbpp_utxos || utxo.nonRgbpp) &&
+          (!only_confirmed || utxo.status.confirmed) &&
+          (!min_satoshi || utxo.value >= min_satoshi),
+      ),
+    );
+  }
+
+  /*
+   * The following methods are not available in offline mode.
+   */
+
+  getBtcRecommendedFeeRates() {
```

Comment:
Can fee rates be pre-defined in `OfflineBtcData`?

Reply:
Sure, we could define a default fee rate to ensure better compatibility with existing implementations. At the same time, we can include comments to encourage users to provide their own real-time fee rates for improved accuracy.

Reply:
Please refer to commit 2b1ccb177ce734fa3d52e35a353da54ca0465480 for the changes.